src/runtime_src/core/tools/xbflash2/OO_Program_Spi.cpp

```diff
@@ -0,0 +1,220 @@
+
+// 3rd Party Library - Include Files
+#include <boost/program_options.hpp>
+#include <boost/format.hpp>
+#include <boost/algorithm/string.hpp>
+namespace po = boost::program_options;
+
+// System - Include Files
+#include <iostream>
+#include <fstream>
+
+namespace {
+
+int
+reset(po::variables_map vm, int bar, size_t baroff) {
+    std::string bdf;
+    bool force = false;
+    bool dualflash = false;
+    XBFU::sudo_or_throw();
+    //mandatory command line args
+    try {
+        bdf = vm["device"].as<std::string>();
+    }
+    catch (...) {
+        return -EINVAL;
+    }
+
+    //optional command line args
+    try {
+        if (vm.count("force"))
+            force = true;
+        if (vm.count("dual-flash"))
+            dualflash = true;
+    }
+    catch (...) {
+    }
+
+    std::cout << "About to revert to golden image for device " << bdf << std::endl;
+
+    if (!force && !XBFU::can_proceed())
+        return -ECANCELED;
+
+    pcidev::pci_device dev(bdf, bar, baroff);
+    XSPI_Flasher xspi(&dev, dualflash);
+    return xspi.revertToMFG();
+}
+
+int
+flash(po::variables_map vm, int bar, size_t baroff) {
+    std::string bdf;
+    std::vector <std::string> primary_file;
+    int ret = 0;
+    bool force = false;
+    bool dual_flash = false;
+
+    XBFU::sudo_or_throw();
+
+    //mandatory command line args
+    try {
+        bdf = vm["device"].as<std::string>();
+        primary_file = vm["image"].as<std::vector<std::string>>();
+        if (primary_file.size() == 2)
+            dual_flash = true;
+    }
+    catch (...) {
+        return -EINVAL;
+    }
+
+    if (vm.count("force"))
+        force = true;
+
+    std::cout
+        << "About to flash below MCS bitstream onto device "
+        << bdf << ":" << std::endl;
+
+    if (!force && !XBFU::can_proceed())
+        return -ECANCELED;
+
+    pcidev::pci_device dev(bdf, bar, baroff);
+    XSPI_Flasher xspi(&dev, dual_flash);
+
+    if (!dual_flash) {
+        firmwareImage pri(primary_file[0].c_str());
+        if (pri.fail())
+            return -EINVAL;
+        ret = xspi.xclUpgradeFirmware1(pri);
+    }
+    else {
+        firmwareImage pri(primary_file[0].c_str());
+        firmwareImage sec(primary_file[1].c_str());
+        if (pri.fail() || sec.fail())
+            return -EINVAL;
+        ret = xspi.xclUpgradeFirmware2(pri, sec);
+    }
+    return ret;
+}
+
+int
+spiCommand(po::variables_map vm) {
+    std::string sBar, sBarOffset;
+    size_t baroff = INVALID_OFFSET;
+    int bar = 0;
+    int err = -EINVAL;
+
+    //optional command line args
+    try {
+        sBar = vm["bar"].as<std::string>();
+        if (!sBar.empty())
+            bar = std::stoi(sBar);
+        sBarOffset = vm["bar-offset"].as<std::string>();
+        if (!sBarOffset.empty()) {
+            std::stringstream sstream(sBarOffset);
+            sstream >> baroff;
+        }
+    }
+    catch (...) {
+    }
+
+    if (vm.count("revert-to-golden")) {
+        err = reset(vm, bar, baroff);
```

Comment:
why are "bar" and "baroff" being passed as arguments if they can be read from vm which is also passed in as an argument?

Reply:
comments addressed.